docroot/modules/custom/mass_entity_usage/mass_entity_usage.module

```diff
@@ -0,0 +1,116 @@
+<?php
+
+/**
+ * @file
+ * Contains mass_entity_usage.module.
+ */
+
+use Drupal\content_moderation\Entity\ContentModerationState;
+use Drupal\content_moderation\Entity\ContentModerationStateInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\mass_content_moderation\MassModeration;
+use Drupal\mayflower\Helper;
+
+/**
+ * Implements hook_entity_usage_block_tracking().
+ */
+function mass_entity_usage_entity_usage_block_tracking($target_id, $target_type, $source_id, $source_type, $source_langcode, $source_vid, $method, $field_name, $count) {
+  // Block tracking of entities with a node source that is unpublished or trash.
+  if (!_mass_entity_usage_tracking_check($source_type, $source_id, $source_vid)) {
+    return TRUE;
+  }
+  return FALSE;
+}
+
+/**
+ * Find an id/revision_id record in the default field data table.
+ *
+ * @param string $entity_type
+ *   The entity type.
+ * @param int $id
+ *   The entity id.
+ * @param int $vid
+ *   The entity revision id.
+ *
+ * @return bool
+ *   If true, the content should be tracked. If false, do not track.
+ */
+function _mass_entity_usage_tracking_check($entity_type, $id, $vid) {
+  if ($entity_type == 'paragraph') {
```

Comment:
We generally recommend defaulting new code to an object-oriented style rather than a procedural style. Ideally, a custom module file just contains hook implementations that delegate to services and classes. It's less obvious in this example where there is only one function in the module file. Usually, by the time a module file becomes too large to be easy to maintain and understand, there also isn't the appetite for a major refactoring.

We'd usually use [Typed Entity](https://www.lullabot.com/articles/drupals-bundle-classes-empower-better-code) to allow us to encapsulate the business logic. That also makes it easier for new developers to learn what each bundle type can do, without having to search through many module files. See the article for details on the limitations of core's bundle classes. With that architecture, we'd end up with something like:

1. A new interface such as `\Drupal\mass_entity_usage\TrackableEntityInterface`:
  - It would contain one method, just `isTrackable(): bool`.
2. We would implement the interface on the classes for each bundle.
  - The paragraph block and the node block in the current if statement would move to each of those classes.
3. `\mass_entity_usage_entity_usage_block_tracking` would then be:

```php
$entity = $entityStorageManager->loadRevision($vid);
if ($entity) {
  $wrapped = $repository_manager->wrap($entity);
  if ($wrapped instanceof TrackableEntityInterface) {
    return $wrapped->isTrackable();
  }
}

return FALSE;
```

There's similar ways to do this without bundle classes or typed entities too. But this is a big topic, so I'll stop here for now!

Reply:
Thanks for the advice and the article @deviantintegral! Much appreciated.